### ISSUE TITLE
Which crc32 implementation is this producing?

### ISSUE THREAD

**Issue by xanderdunn** [komrad36/CRC#2]
I find that the results of `option_13_golden_intel` don't match the crc32 values I'm expecting. The results match neither the [crc32c library](https://github.com/htot/crc32c), nor zlib's `crc32()`. 

An example test string: `hello-world` tested between zlib, crc32c, and this package:
```
truth: 2983461467, zlib: 2983461467, crc32c: 4099351003, crc: 1585878819
```

A related discussion is at htot/crc32c#3. Do you know exactly which implementation of crc32 this library is producing?

**Comment by komrad36** [comment]
The README goes over some of the many variations that can cause CRC implementations to differ. Of interest are:

- choice of polynomial
- input reflected?
- output reflected?
- input XOR'd with an initial value (usually 0xFFFFFFFF)?
- output XOR'd with an initial value (usually 0xFFFFFFFF)?

There is also a common ambiguity in the reflectedness of the polynomial.

By tuning these parameters, I can produce each of the 3 outputs you gave in your example.

You can use this site to easily test various permutations:

[http://www.sunshine2k.de/coding/javascript/crc/crc_js.html](http://www.sunshine2k.de/coding/javascript/crc/crc_js.html)

With the following settings for all test cases:

CRC width: CRC-32
CRC parametrization: custom
CRC input data: string, with value "hello-world"

Now:

# To match zlib: #
Input reflected: YES
Output reflected: YES
Polynomial: 0x4C11DB7
Initial value: 0xFFFFFFFF
Final xor value: 0xFFFFFFFF

# To match crc32c: #
Input reflected: YES
Output reflected: YES
Polynomial: 0x1EDC6F41
Initial value: 0xFFFFFFFF
Final xor value: 0xFFFFFFFF

# To match me: #
Input reflected: YES
Output reflected: YES
Polynomial: 0x1EDC6F41
Initial value: 0x0
Final xor value: 0x0

- - -

So the only difference between zlib and CRC32-C is that zlib uses a different polynomial.

The only difference between my implementation and CRC32-C is that my implementation (*by default*) does not XOR the input or output with anything, but this is *configurable*. If you want my implementation to match CRC32-C, check out commented lines [142](https://github.com/komrad36/CRC/blob/master/CRC/golden_intel.cpp#L142) and [176](https://github.com/komrad36/CRC/blob/master/CRC/golden_intel.cpp#L176) in golden-intel.cpp. If you uncomment those (and comment out the currently active alternatives below them), it will perform the input and output XOR and then match CRC32-C.

**Comment by xanderdunn** [comment]
Thanks a lot @komrad36, that's very helpful. I'm actually looking to reproduce zlib, which is used by our data provider. I believe uncommenting the two lines you mentioned above in combination with setting golen_intel.cpp's `constexpr uint32_t P = 0x4C11DB7;` would achieve this?

**Comment by komrad36** [comment]
No problem!

The superfast golden code (Option 13) utilize x86 hardware CRC instructions to go faster than everybody else. (As do options 11 and 12). But those x86 instructions **hardcode** the polynomial and input/output reflectivity. Zlib matches on reflectivity, but uses a different polynomial, so unfortunately, it will **not** be possible to modify the golden code to match zlib's polynomial.

If you must match it, the fastest approach that permits arbitrary choice of polynomial is Option 10 (16-byte tabular), which will be about 8x slower than Option 13, but should still be better than zlib - or perhaps comparable, if zlib is really well optimized (haven't tested).

If you'd like to go that route, you may change the value of [P in tabular_methods.cpp](https://github.com/komrad36/CRC/blob/master/CRC/tabular_methods.cpp#L4), then run with [this set to true](https://github.com/komrad36/CRC/blob/master/CRC/main.cpp#L6) to reprint the tables with the new polynomial. Replace the old tables with the new ones, and you should be good to go with Option 10 (16-byte tabular), which should run at about 8 bits/cycle.

Reopening, let me know how it goes!

**Comment by komrad36** [comment]
To clear up possibly confusion due to the aforementioned ambiguity in reflectedness of polynomial: to match Zlib, you will want the following polynomial:

```cpp
static constexpr uint32_t P = 0xEDB88320U;
```

Which is the reflection of `0x04C11DB7U`.

Also Option 10 should be modified (it was just intended as a demo) to XOR/invert input/output, like Zlib, and also to handle lengths that are not a multiple of 16:

```cpp
uint32_t option_10_tabular_16_bytes(const void* M, uint32_t bytes)
{
    const uint32_t* M32 = (const uint32_t*)M;
    uint32_t R = ~0U;

    for (; bytes >= 16; bytes -= 16)
    {
        R ^= *M32++;
        const uint32_t R2 = *M32++;
        const uint32_t R3 = *M32++;
        const uint32_t R4 = *M32++;
        R = g_tbl[ 0 * 256 + uint8_t(R4 >> 24)] ^
            g_tbl[ 1 * 256 + uint8_t(R4 >> 16)] ^
            g_tbl[ 2 * 256 + uint8_t(R4 >> 8)] ^
            g_tbl[ 3 * 256 + uint8_t(R4 >> 0)] ^
            g_tbl[ 4 * 256 + uint8_t(R3 >> 24)] ^
            g_tbl[ 5 * 256 + uint8_t(R3 >> 16)] ^
            g_tbl[ 6 * 256 + uint8_t(R3 >> 8)] ^
            g_tbl[ 7 * 256 + uint8_t(R3 >> 0)] ^
            g_tbl[ 8 * 256 + uint8_t(R2 >> 24)] ^
            g_tbl[ 9 * 256 + uint8_t(R2 >> 16)] ^
            g_tbl[10 * 256 + uint8_t(R2 >> 8)] ^
            g_tbl[11 * 256 + uint8_t(R2 >> 0)] ^
            g_tbl[12 * 256 + uint8_t( R >> 24)] ^
            g_tbl[13 * 256 + uint8_t( R >> 16)] ^
            g_tbl[14 * 256 + uint8_t( R >> 8)] ^
            g_tbl[15 * 256 + uint8_t( R >> 0)];
    }

    const uint8_t* M8 = (const uint8_t*)M32;

    for (; bytes; --bytes)
    {
        R = (R >> 8) ^ g_tbl[uint8_t(R ^ *M8++)];
    }

    return ~R;
}
```

This will match Zlib once the tables have been rebuilt using the above value for P, e.g. it produces `2983461467` for input string `hello-world` and length 11.

**Comment by komrad36** [comment]
I've added comments to the sources indicating which polynomials can safely be changed

**Comment by xanderdunn** [comment]
Thanks @komrad36. I was able to print the table values and include them in tabular_methods.cpp, but then I got a seg fault trying to run it. This is probably because of the necessary modifications you mentioned above. At the moment profiling shows our application's bottleneck isn't the crc32 hashing, so I won't spend more time on this right now. Feel free to close the issue, I'll come back if it becomes a priority again.
